File: DL-PM/DL-PM.cs

/* https://www.decentlab.com/products/particulate-matter-temperature-humidity-and-barometric-pressure-sensor-for-lorawan */

using System;
using System.IO;
using System.Collections.Generic;

public class DecentlabDecoder
{
  private delegate double conversion(double[] x);
  public const int PROTOCOL_VERSION = 2;

  private class Sensor
  {
    internal int length { get; set; }
    internal List<SensorValue> values { get; set; }
    internal Sensor(int length, List<SensorValue> values)
    {
      this.length = length;
      this.values = values;
    }
  }

  private class SensorValue
  {
    internal string name { get; set; }
    internal string unit { get; set; }
    internal conversion convert;
    internal SensorValue(string name, string unit, conversion convert)
    {
      this.name = name;
      this.unit = unit;
      this.convert = convert;
    }
  }

  private static readonly List<Sensor> SENSORS = new List<Sensor>() {
    new Sensor(1, new List<SensorValue>() {
      new SensorValue("Battery voltage", "V", x => x[0] / 1000)
    }),
    new Sensor(10, new List<SensorValue>() {
      new SensorValue("PM1.0 mass concentration", "µg⋅m⁻³", x => x[0] / 10),
      new SensorValue("PM2.5 mass concentration", "µg⋅m⁻³", x => x[1] / 10),
      new SensorValue("PM4 mass concentration", "µg⋅m⁻³", x => x[2] / 10),
      new SensorValue("PM10 mass concentration", "µg⋅m⁻³", x => x[3] / 10),
      new SensorValue("Typical particle size", "nm", x => x[4]),
      new SensorValue("PM0.5 number concentration", null, x => x[5] / 10),
      new SensorValue("PM1.0 number concentration", null, x => x[6] / 10),
      new SensorValue("PM2.5 number concentration", null, x => x[7] / 10),
      new SensorValue("PM4 number concentration", null, x => x[8] / 10),
      new SensorValue("PM10 number concentration", null, x => x[9] / 10)
    }),
    new Sensor(2, new List<SensorValue>() {
      new SensorValue("Air temperature", "°C", x => 175.72 * x[0] / 65536 - 46.85),
      new SensorValue("Air humidity", "%", x => 125 * x[1] / 65536 - 6)
    }),
    new Sensor(1, new List<SensorValue>() {
      new SensorValue("Barometric pressure", "Pa", x => x[0] * 2)
    })
  };

  private static int ReadInt(Stream stream)
  {
    return (stream.ReadByte() << 8) + stream.ReadByte();
  }


  public static Dictionary<string, Tuple<double, string>> Decode(byte[] msg)
  {
    return Decode(new MemoryStream(msg));
  }

  public static Dictionary<string, Tuple<double, string>> Decode(String msg)
  {
    byte[] output = new byte[msg.Length / 2];
    for (int i = 0, j = 0; i < msg.Length; i += 2, j++)
    {
      output[j] = (byte)int.Parse(msg.Substring(i, 2), System.Globalization.NumberStyles.HexNumber);
    }
    return Decode(output);
  }

  public static Dictionary<string, Tuple<double, string>> Decode(Stream msg)
  {
    var version = msg.ReadByte();
    if (version != PROTOCOL_VERSION)
    {
      throw new InvalidDataException("protocol version " + version + " doesn't match v2");
    }

    var result = new Dictionary<string, Tuple<double, string>>();
    result["Protocol version"] = new Tuple<double, string>(version, null);

    var deviceId = ReadInt(msg);
    result["Device ID"] = new Tuple<double, string>(deviceId, null);

    var flags = ReadInt(msg);
    foreach (var sensor in SENSORS)
    {
      if ((flags & 1) == 1)
      {
        double[] x = new double[sensor.length];
        for (int i = 0; i < sensor.length; i++)
        {
          x[i] = ReadInt(msg);
        }
        foreach (var val in sensor.values)
        {
          if (val.convert != null)
          {
            result[val.name] = new Tuple<double, string>(val.convert(x), val.unit);
          }
        }
      }
      flags >>= 1;
    }

    return result;
  }
}
public class Program
{
  public static void Main()
  {
    var payloads = new string[] {
      "021b50000f0c25002500270027002701f50107012c012d012d012d67bd618dbd10",
      "021b50000d0c2567bd618dbd10",
      "021b5000010c25"
    };

    foreach (var pl in payloads)
    {
      var decoded = DecentlabDecoder.Decode(pl);
      foreach (var k in decoded.Keys)
      {
        Console.WriteLine(k + ": " + decoded[k]);
      }
      Console.WriteLine();
    }
  }
}

File: decoders/DL-PM/DL-PM.erl

% https://www.decentlab.com/products/particulate-matter-temperature-humidity-and-barometric-pressure-sensor-for-lorawan

-module(decentlab_decoder).
-define(PROTOCOL_VERSION, 2).
-export([decode/1, start/0]).

-define(SENSOR_DEFS,
  [
    #{
      length => 1,
      values => [
        #{
          name => <<"Battery voltage">>,
          convert => fun(X) -> lists:nth(0 + 1, X) / 1000 end,
          unit => <<"V"/utf8>>
        }
      ]
    },
    #{
      length => 10,
      values => [
        #{
          name => <<"PM1.0 mass concentration">>,
          convert => fun(X) -> lists:nth(0 + 1, X) / 10 end,
          unit => <<"µg⋅m⁻³"/utf8>>
        },
        #{
          name => <<"PM2.5 mass concentration">>,
          convert => fun(X) -> lists:nth(1 + 1, X) / 10 end,
          unit => <<"µg⋅m⁻³"/utf8>>
        },
        #{
          name => <<"PM4 mass concentration">>,
          convert => fun(X) -> lists:nth(2 + 1, X) / 10 end,
          unit => <<"µg⋅m⁻³"/utf8>>
        },
        #{
          name => <<"PM10 mass concentration">>,
          convert => fun(X) -> lists:nth(3 + 1, X) / 10 end,
          unit => <<"µg⋅m⁻³"/utf8>>
        },
        #{
          name => <<"Typical particle size">>,
          convert => fun(X) -> lists:nth(4 + 1, X) end,
          unit => <<"nm"/utf8>>
        },
        #{
          name => <<"PM0.5 number concentration">>,
          convert => fun(X) -> lists:nth(5 + 1, X) / 10 end,
          unit => <<"1⋅cm⁻³"/utf8>>
        },
        #{
          name => <<"PM1.0 number concentration">>,
          convert => fun(X) -> lists:nth(6 + 1, X) / 10 end,
          unit => <<"1⋅cm⁻³"/utf8>>
        },
        #{
          name => <<"PM2.5 number concentration">>,
          convert => fun(X) -> lists:nth(7 + 1, X) / 10 end,
          unit => <<"1⋅cm⁻³"/utf8>>
        },
        #{
          name => <<"PM4 number concentration">>,
          convert => fun(X) -> lists:nth(8 + 1, X) / 10 end,
          unit => <<"1⋅cm⁻³"/utf8>>
        },
        #{
          name => <<"PM10 number concentration">>,
          convert => fun(X) -> lists:nth(9 + 1, X) / 10 end,
          unit => <<"1⋅cm⁻³"/utf8>>
        }
      ]
    },
    #{
      length => 2,
      values => [
        #{
          name => <<"Air temperature">>,
          convert => fun(X) -> 175.72 * lists:nth(0 + 1, X) / 65536 - 46.85 end,
          unit => <<"°C"/utf8>>
        },
        #{
          name => <<"Air humidity">>,
          convert => fun(X) -> 125 * lists:nth(1 + 1, X) / 65536 - 6 end,
          unit => <<"%"/utf8>>
        }
      ]
    },
    #{
      length => 1,
      values => [
        #{
          name => <<"Barometric pressure">>,
          convert => fun(X) -> lists:nth(0 + 1, X) * 2 end,
          unit => <<"Pa"/utf8>>
        }
      ]
    }
  ]
).


decode(<<?PROTOCOL_VERSION, DeviceId:16,
         Flags:16/bitstring, Bytes/binary>> = Binary) when is_binary(Binary) ->
  Words = bytes_to_words(Bytes),
  maps:merge(#{<<"Protocol version">> => ?PROTOCOL_VERSION,
               <<"Device ID">> => DeviceId},
             sensor(Words, Flags, ?SENSOR_DEFS));
decode(HexStr) ->
  Binary = hex2bin(HexStr),
  decode(Binary).


start() ->
  io:format("~p~n", [decode("021b50000f0c25002500270027002701f50107012c012d012d012d67bd618dbd10")]),
  io:format("~p~n", [decode("021b50000d0c2567bd618dbd10")]),
  io:format("~p~n", [decode("021b5000010c25")]).


%%%%%%%%%%%%%%%%%%%%%%%%%%%%% PRIVATE FUNCTIONS %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

sensor(_Words, _Flags, []) -> #{};
sensor(Words, <<Flags:15, 0:1>>, [_Cur | Rest]) ->
  sensor(Words, <<0:1, Flags:15>>, Rest);
sensor(Words, <<Flags:15, 1:1>>,
       [#{length := Len, values := ValueDefs} | RestDefs]) ->
  {X, RestWords} = lists:split(Len, Words),
  Values = value(ValueDefs, X),
  RestSensors = sensor(RestWords, <<0:1, Flags:15>>, RestDefs),
  maps:merge(Values, RestSensors).


value([], _X) -> #{};
value([#{convert := Convert,
         name := Name} = ValueDef | RestDefs], X) ->
  Values = value(RestDefs, X),
  Value = #{<<"value">> => Convert(X)},
  Value2 = add_unit(Value, ValueDef),
  maps:put(Name, Value2, Values);
value([_NoConvert | RestDefs], X) -> value(RestDefs, X).


add_unit(Value, #{unit := Unit}) ->
  maps:put(<<"unit">>, Unit, Value);
add_unit(Value, _NoUnit) ->
  Value.


hex2bin(HexStr) when is_binary(HexStr) -> hex2bin(binary_to_list(HexStr));
hex2bin(HexStr) -> hex2bin(HexStr, []).

hex2bin("", Bins) -> list_to_binary(lists:reverse(Bins));
hex2bin(HexStr, Bins) ->
  Bytes = string:slice(HexStr, 0, 2),
  Word = list_to_integer(Bytes, 16),
  hex2bin(string:slice(HexStr, 2), [Word | Bins]).


bytes_to_words(Bytes) -> bytes_to_words(Bytes, []).

bytes_to_words(<<>>, Words) -> lists:reverse(Words);
bytes_to_words(<<Word:16, Rest/binary>>, Words) ->
  bytes_to_words(Rest, [Word | Words]).

where(true, IfTrue, _IfFalse) -> IfTrue;
where(false, _IfTrue, IfFalse) -> IfFalse.